### PR TITLE
Add Source 2 Typed Variant

### PR DESCRIPTION
**Pull request by zhaojj2209**
Changes:
-
- Added support for `null` to type error checker.
- Added `Pair` and `List` types to type error checker.
- Added type checking for Source 2 library functions (e.g. `pair`, `list`, `head`, `tail`).
- Added tests for Source 2 Typed.

How to Test this PR:
-
1. Without frontend: Run Source 2 Typed in the js-slang REPL.
```
$ js-slang -c 2 -v typed # if js-slang is added to PATH
$ node dist/repl/repl.js -c 2 -v typed # if js-slang is not added to PATH
```
2. With frontend: [Link local js-slang to local frontend](https://github.com/source-academy/js-slang#using-your-js-slang-in-your-local-source-academy) and checkout local frontend to source-academy/frontend#2293.
3. Check the existing unit tests in `src/typeChecker/__tests__/source2Typed.test.ts`.
```
$ yarn test src/typeChecker/__tests__/source2Typed.test.ts
```

### CODE REVIEW

File: src/typeChecker/typeErrorChecker.ts [hunk 1]
```diff
@@ -251,7 +257,61 @@ function typeCheckAndReturnType(node: tsEs.Node, context: Context, env: TypeEnvi
...
+          checkForTypeMismatch(node, actualType, expectedType, context)
+          if (context.errors.length > numErrors) {
+            // If errors were found, return any type
+            return tAny
+          }
+          if (actualType.kind === 'pair') {
+            return fnName === 'head' ? actualType.headType : actualType.tailType
+          }
+          if (actualType.kind === 'list') {
+            return fnName === 'head' ? actualType.elementType : actualType
```
Comment (ianyong) [inline src/typeChecker/typeErrorChecker.ts:309]: Given that `List` is defined as such:
```ts
export interface List {
  kind: 'list'
  elementType: Type
  // Used in Source Typed variants to check for type mismatches against pairs
  typeAsPair?: Pair
}
```
If we take the `tail` of a list, the `typeAsPair` remains unchanged despite the list itself changing. Is this intended?

File: src/typeChecker/typeErrorChecker.ts [hunk 2]
```diff
@@ -397,8 +456,21 @@ function addTypeDeclarationsToEnvironment(
         setDeclKind(id.name, node.kind, env)
         break
       case 'TSTypeAliasDeclaration':
+        const alias = node.id.name
+        if (context.chapter >= 2 && (alias === 'Pair' || alias === 'List')) {
+          throw new TypecheckError(
+            node,
+            `Type '${alias}' is a reserved type and cannot be redeclared`
+          )
+        }
```
Comment (ianyong) [inline src/typeChecker/typeErrorChecker.ts:465]: After looking at this, I realised that I can do things like:
```ts
type string = number;
```

For reference, TypeScript disallows this:
```
TS2457: Type alias name cannot be 'string'.
```

File: src/typeChecker/utils.ts [hunk 3]
```diff
@@ -352,16 +362,41 @@ export const temporaryStreamFuncs: [string, BindableType][] = [
...
+  ['member', tFunc(tAny, tList(tAny), tList(tAny))],
+  ['remove', tFunc(tAny, tList(tAny), tList(tAny))],
+  ['remove_all', tFunc(tAny, tList(tAny), tList(tAny))],
+  ['reverse', tFunc(tList(tAny), tList(tAny))],
+  // misc functions
+  // TODO: Add support for type checking of functions with variable no. of args
+  ['display_list', tForAll(tAny)],
+  ['draw_data', tForAll(tAny)],
+  ['equal', tFunc(tAny, tAny, tBool)]
 ]
```
Comment (ianyong) [inline src/typeChecker/utils.ts:400]: To confirm, this means that once I run any kind of operation on lists, I lose the type of its elements?

File: src/typeChecker/__tests__/source2Typed.test.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    `
+
+    parse(code, context)
+    expect(parseError(context.errors)).toMatchInlineSnapshot(`
+      "Line 2: Expected 2 arguments, but got 1.
+      Line 3: Expected 2 arguments, but got 3."
+    `)
+  })
+
+  it('lists are pairs', () => {
```
Comment (ianyong) [inline src/typeChecker/__tests__/source2Typed.test.ts:61]: Is this always true? What about the empty list?

File: src/typeChecker/__tests__/source2Typed.test.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+describe('list', () => {
+  it('handles type mismatches correctly', () => {
+    const code = `const x1: List<number> = list(1, 2, 3); // no error
+      const x2: List<number> = list(); // no error
+      const x3: List<number> = list('1'); // error
+      const x4: List<number> = 1; // error
+      const x5: List<number> = list(1, '2'); // error
+      const x6: List<number | string> = list(1, '2'); // no error
+      const x7: List<number, string> = list(1, '2'); // error
+    `
```
Comment (ianyong) [inline src/typeChecker/__tests__/source2Typed.test.ts:80]: Maybe can add a test case for a list of `null`s for completeness?

File: src/typeChecker/typeErrorChecker.ts (hunk 1 above)
Comment (ianyong) [reply src/typeChecker/typeErrorChecker.ts:309]: Would also be good to add a test case that can catch this issue as a form of regression testing.

**Review by ianyong — CHANGES_REQUESTED** [review]

File: src/errors/typeErrors.ts [hunk 6]
```diff
@@ -593,3 +593,23 @@ export class InvalidNumberOfArgumentsTypeError implements SourceError {
     return `Try calling function ${calleeStr} again, but with ${this.expected} argument${pluralS} instead. Remember that arguments are separated by a ',' (comma).`
   }
 }
+
+export class IncorrectNumberOfTypeArgumentsError implements SourceError {
```
[the remark this answers is not in the document]
Comment (zhaojj2209) [reply src/errors/typeErrors.ts:597]: `InvalidNumberOfArguments` is for arguments to function calls, while `IncorrectNumberOfTypeArgumentsError` is for type parameters to generic types (e.g. `Pair`, `List`). Will rename this error for clarity.

File: src/typeChecker/typeErrorChecker.ts (hunk 1 above)
Comment (zhaojj2209) [reply src/typeChecker/typeErrorChecker.ts:309]: Thanks for the catch! Will make the necessary changes.

File: src/typeChecker/typeErrorChecker.ts (hunk 2 above)
Comment (zhaojj2209) [reply src/typeChecker/typeErrorChecker.ts:465]: Got it, will follow TS and disallow overwriting of basic types using type aliases.

File: src/typeChecker/utils.ts (hunk 3 above)
Comment (zhaojj2209) [reply src/typeChecker/utils.ts:400]: Unfortunately as of right now the type error checker is not yet powerful enough to infer the return type based on the types of its arguments. I will KIV this for future enhancements.

File: src/typeChecker/__tests__/source2Typed.test.ts (hunk 4 above)
Comment (zhaojj2209) [reply src/typeChecker/__tests__/source2Typed.test.ts:61]: I don't think `null` is considered a pair, given that the definition of a list is "either `null` or a pair whose tail is a list". Will rename this for clarity and also add testcases for the case where `null` should not be typed as `Pair`.

**Review by ianyong — APPROVED** [review]
